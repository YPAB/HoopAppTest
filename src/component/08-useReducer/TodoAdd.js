import React from "react";
import { useForm } from "../../hooks/useForm";
import './styles.css'

export const TodoAdd = ({ handleAddTask }) => {

    const [{description},handleInputChange,reset ] = useForm({
        description:''

    });

    const handleSubmit = (e) =>{
        e.preventDefault();

        if (description.trim().length <= 1 ) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc:description,
            done: false
        };

        handleAddTask( newTodo );
        reset();
        
    }

    return (
        <>
             <h4> Agregar Tarea</h4>
                   <hr/>

                   <form onSubmit={ handleSubmit }>
                       <input
                        type='text'
                        name='description'
                        className='form-control'
                        placeholder='Aprender ...'
                        autoComplete='off'
                        value={ description }
                        onChange={ handleInputChange }
                       />

                       <button
                        type='submit'
                        className='btn btn-primary mt-1 btn-blck'
                       >
                        Agregar
                       </button>

                    </form>
        </>


    );

}