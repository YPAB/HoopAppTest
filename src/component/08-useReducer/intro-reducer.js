const tasks = [{
    id:1,
    todo:'Comprar Pan',
    done:false
}];

const todoReducer = ( state = tasks,action) => {

    if ( action?.type == 'create' ) {
        return [ ...state, action.payload ];
    }

    return state;
}

let todos = todoReducer();

const newTodo = {
    id:2,
    todo:'Comprar Noot',
    done:false
}


const agregarTodoAction = {
    type:'create',
    payload: newTodo
}

todos = todoReducer( todos, agregarTodoAction );

console.log( todos );