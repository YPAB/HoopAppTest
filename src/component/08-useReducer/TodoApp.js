import React,{useEffect, useReducer}from 'react';
import './styles.css'
import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

export const TodoApp = () => {

    const init = () => {


        //si hay todos traelos o sino retorna un arreglo vacio
        return JSON.parse(localStorage.getItem('todos')) || [];


        /* return [{
            id: new Date().getTime(),
            desc:'Aprender React',
            done: false
        }]; */


    }

    /* const initialState = [ {
        id: new Date().getTime(),
        desc:'Aprender React',
        done: false
    }]; */

        //el dispath es una funcion que le mandamos una accion y ya va a saber a que reducer es el que hay que mandar.
    const [ todos,dispatch ] = useReducer( todoReducer,[], init );
    

    //console.log(todos);

    useEffect ( () => {
        localStorage.setItem('todos',JSON.stringify( todos ))

    },[todos]);

    const handleToggle = (todoId) => {

        const action = {
            type: 'toggle',
            payload: todoId
        }

        dispatch(action);

    }

    const handleAddTask = ( newTodo ) => {
        dispatch ( {
            type:'create',
            payload:newTodo
        });
    }


    const handleDelete = ( todoId ) => {
        const action = {
            type: 'delete',
            payload: todoId

        }

        dispatch(action);
    }
 
    return (
        <div>
            <h1>Tareas App: { todos.length }</h1>
            <hr/>

           <div className='row'>
               <div className='col-7'>
                   <TodoList 
                        todos = {todos} 
                        handleDelete= {handleDelete} 
                        handleToggle= { handleToggle} 
                    />
               </div>

               <div className='col-5'>
                  <TodoAdd
                    handleAddTask = { handleAddTask }
                  />

               </div>

           </div>

        </div>
    )
}