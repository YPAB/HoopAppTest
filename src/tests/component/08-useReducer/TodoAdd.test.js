import '@testing-library/jest-dom';
import { TodoAdd } from '../../../component/08-useReducer/TodoAdd'
import { shallow } from 'enzyme';
import { demoTodos } from '../../fixtures/demoTodos';

describe('Pruebas en <TodoAdd></TodoAdd>', () => {

    const handleAddTask = jest.fn();

    const wrapper = shallow(<TodoAdd
        handleAddTask= { handleAddTask }
    />)

    test('debe de mostrarse correctamente', () => {

        expect( wrapper ).toMatchSnapshot();
    });

    test('NO debe de llamar a handleAddTodo', () => {

        const formSubmit = wrapper.find('form').prop('onSubmit');

        formSubmit({preventDefault() {} });

        expect( handleAddTask ).toHaveBeenCalledTimes(0);
        

    });


    test('Debe de llamar la funcion handleAddTodo', () => {

        const value = 'Aprender React';

        wrapper.find('input').simulate('change', {
            target: {
                value,
                name: 'description'
            }
        });

        const formSubmit = wrapper.find('form').prop('onSubmit');

        formSubmit({preventDefault() {} });

        expect( handleAddTask ).toHaveBeenCalledTimes(1);
        expect( handleAddTask ).toHaveBeenCalledWith( {
            id: expect.any(Number),
            desc: value,
            done: false
        });

        expect( wrapper.find('input').prop('value') ).toBe('');

        

    });





})