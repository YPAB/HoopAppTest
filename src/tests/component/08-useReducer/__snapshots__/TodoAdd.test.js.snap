// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoAdd></TodoAdd> debe de mostrarse correctamente 1`] = `
<Fragment>
  <h4>
     Agregar Tarea
  </h4>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="form-control"
      name="description"
      onChange={[Function]}
      placeholder="Aprender ..."
      type="text"
      value=""
    />
    <button
      className="btn btn-primary mt-1 btn-blck"
      type="submit"
    >
      Agregar
    </button>
  </form>
</Fragment>
`;
